### PR TITLE
[ENH] GRU-based time series classifiers

### PR DESCRIPTION
**Pull request by fnhirwa**
This PR adds the implementation for GRU-based classifiers.

The implementation fixes an issue with the `PyTorchDataset` custom class that was returning an array of inputs to be passed to the model and modified to return a dict.

Part of https://github.com/sktime/sktime/issues/6660

## Implemented models

- [x] GRU (Cho, 2014)  [Paper](https://arxiv.org/abs/1412.3555)
- [x] GRU-FCN(Elsayed, 2018)[Paper](https://arxiv.org/abs/1812.07683)

### CODE REVIEW

File: sktime/classification/deep_learning/gru.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+
+from sktime.classification.deep_learning._pytorch import BaseDeepClassifierPytorch
+from sktime.utils.dependencies import _check_soft_dependencies
+
+if _check_soft_dependencies("torch", severity="none"):
+    import torch
+
+    from sktime.networks.gru import GRU, GRUFCNN
+
+    OPTIMIZERS = {
```
Comment (fkiraly) [inline sktime/classification/deep_learning/gru.py:13]: should this be in a more central location perhaps? These are very general, so could be put as a (private) method into the torch base class.

Comment (fnhirwa) [reply sktime/classification/deep_learning/gru.py:13]: Yeah, this can be moved to the base class.

**Comment by fnhirwa** [comment]
> Looks reasonable to me, I have not reviewed how the networks map on what is describe din the papers, but I assume you checked that.
> 
> I also assume there is code you copied for this? Please credit the authors of that code by (a) referencing the original location and (b) adding them to the `authors` tag.
> 
> Further, the networks are plain objects in `networks` - which I suppose is fine (no blocker), we may consider refactoring later though, compare the dedicated network classes for the `tensorflow` objects. Not sure which option is better though.

The implementations I took as references were tensorflow based and I properly added the references in the docstring. Regarding the refactor, we need to thoroughly refactor and dedicate the objects based on the underlying frameworks.

**Comment by fkiraly** [comment]
what is going on with the failing env in `test-unix-pandas1`?

**Comment by fnhirwa** [comment]
> what is going on with the failing env in `test-unix-pandas1`?

It seems like a dependency issue when the runner tries to run that.
I can see an issue with `dash` conflicting with `Werkzeug`. Should we add strict requirements to the `project.toml` for this? 

https://github.com/sktime/sktime/blob/9f214e5a65a4cf2f09ef8dd469a798cbf60cb011/pyproject.toml#L83

**Comment by fkiraly** [comment]
We don't seem to be changing pyproject.toml - is this new?

**Comment by fnhirwa** [comment]
> We don't seem to be changing pyproject.toml - is this new?

I didn't make any changes on the `pyproject.toml` but I think the issue for some runners is non-determinism in choosing which version of `dash` to install that's why the failure is random for this case.

**Comment by fkiraly** [comment]
going to restart

**Comment by fnhirwa** [comment]
they are some segmentation fault errors not sure if they are originating from my changes as locally on macos every tests passes.

i’ll try to deep into this and see the cause.

**Comment by fkiraly** [comment]
May I suggest to use this to diagnose which of the two possibilities is true, i.e., from the estimator itself or somewhere else?
https://github.com/sktime/sktime/pull/7087

(add the `test_vm` tag to the estimator, and optionally remove it from the FM to avoid three estimators tested - I suggest a new branch to avoid "contaminating" this one with a merge)

**Pull request merged**